scipy/sparse/linalg/_dsolve/tests/test_linsolve.py

```diff
@@ -722,6 +722,47 @@ def worker():
 
         assert_equal(len(oks), 20)
 
+class TestGstrsErrors:
+    def setup_method(self):
+      self.A = array([[1.0,2.0,3.0],[4.0,5.0,6.0],[7.0,8.0,9.0]], dtype=np.float64)
+      self.b = np.array([[1.0],[2.0],[3.0]], dtype=np.float64)
+
+    def test_trans(self):
+        L = scipy.sparse.tril(self.A, format='csc')
+        U = scipy.sparse.triu(self.A, k=1, format='csc')
+        with assert_raises(ValueError, match="trans must be N, T, or H"):
+            _superlu.gstrs('X', L.shape[0], L.nnz, L.data, L.indices, L.indptr,
+                                U.shape[0], U.nnz, U.data, U.indices, U.indptr, self.b)
```

Comment:
Would these errors not occur if the public `spsolve_triangular` method is called with similar arguments? Usually we test that the functions in the public API throw the correct errors, not the internal functions.

Reply:
The [Python layer of `spsolve_triangular`](https://github.com/maxaehle/scipy/blob/fca2e3b7ec802b0f43be28fa686f9caf1b8dbc01/scipy/sparse/linalg/_dsolve/linsolve.py#L606) ensures that dimensions of the matrix and vector match (by raising an error if not), and that their scalar types agree and are a floating-point type (via type conversions, if necessary). The Python code then calls the new [`gstrs` function in the C layer](https://github.com/maxaehle/scipy/blob/fca2e3b7ec802b0f43be28fa686f9caf1b8dbc01/scipy/sparse/linalg/_dsolve/_superlumodule.c#L267). If the Python code is correct, it should be impossible to trigger the errors in `gstrs` by supplying bad arguments to `spsolve_triangular`. Errors raised in `gstrs` hint to bugs in the Python implementation of `spsolve_triangular`, or any other SciPy code that may use `gstrs` in the future. Therefore, I think it is fine to test these errors using the private API. This is similar to [existent checks of `gssv`](https://github.com/maxaehle/scipy/blob/fca2e3b7ec802b0f43be28fa686f9caf1b8dbc01/scipy/sparse/linalg/_dsolve/tests/test_linsolve.py#L365-L399).